### DISCUSSION TITLE
How does the Datadog Exporter attach metrics to Hosts?

### DISCUSSION THREAD

**Discussion by bertysentry** [open-telemetry/opentelemetry-collector-contrib#6125] [Q&A] [answered]
Say we have the below pipeline:
* Custom Prometheus exporter, exposing custom metrics on /metrics
* OpenTelemetry Collector Prometheus Receiver consuming these metrics
* OpenTelemetry Collector Datadog Exporter pushing these metrics to Datadog

**Problem**: All metrics are attached to the same *Host* (in Datadog's terminology), which is simply the host where our OpenTelemetry Collector is running.

**Expected**: The collected metrics have a *hostname* label, and the corresponding *Host* entries should be listed in Datadog (in Infrastructure, and Host map, etc.), and the corresponding metrics should be associated to it.

The [Datadog Exporter code](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/datadogexporter/internal/translator/metrics_translator.go) shows that it has some logic to translate various hostname labels and convert them as *Hosts*. However, I tried adding various labels to my Prometheus metrics (`host`, `hostname`, `host.name`, `datadog.host.name`, `datadog_host_name`, etc.), and none of them triggered the creation of the corresponding *Hosts*.

So, what labels must be added to the metrics so that the Datadog Exporter creates the corresponding *Hosts*?

Thank you!

**Comment by bertysentry** [comment]
@mx-psi, @ericmustin and @KSerrania any idea? Thank you! 😊

**Reply by ericmustin** [reply]
@bertysentry I'm no longer employed with datadog and defer to @mx-psi and @KSerrania here. Off the top of my head, i would say that you may need to ensure you're operating on [resource attributes](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/cff02ccae545a87cb9fd50049d6c3428b1aebf27/exporter/datadogexporter/internal/translator/metrics_translator.go#L411), but not too sure how prometheus labels get massaged to OTLP attributes or resource attributes so can't say with much confidence whether that's already the case or whether you need to set those details somewhere else on your metrics.

**Reply by bertysentry** [reply]
Thank you @ericmustin! Yeah I think you're getting close to the cause of my problem. Don't worry about it though if you no longer work at Datadog, I just got in touch with other persons who may be able to help. Again, thank you! 😊

**Reply by mx-psi** [reply]
Thanks @ericmustin for the ping :) We will speak in private about this with @bertysentry's team (and I will post something in here for everyone to see at some point, so that people stumbling upon this can get some answers too!)

**Comment by bertysentry** [comment]
Maybe this is the solution: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/6232 (Add support for metrics in [groupbyattrsprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/groupbyattrsprocessor))
;-)

**Comment by bertysentry** [comment, marked as answer]
As @ericmustin mentioned, the Datadog exporter leverages the "hostname", "host.name" (and equivalent) on the *Resource* a *Metric* is attached to, to create the corresponding *Hosts* in Datadog. It doesn't use the attributes of the *Metric* itself (which makes sense).

To solve the problem raised here, #6232 (Add support for metrics in groupbyattrsprocessor) must be implemented so that Prometheus metrics can be attached to *Resources* that are not just the exporter host itself.
